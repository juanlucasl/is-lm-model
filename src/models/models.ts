/* SPDX-License-Identifier: Apache-2.0 */
export type FunctionFormula = (x: number) => number;

export type Variable = {
  value: number;
  setValue: (value: number) => void;
  label: string;
  increaseDescription?: string | JSX.Element;
  decreaseDescription?: string | JSX.Element;
};
