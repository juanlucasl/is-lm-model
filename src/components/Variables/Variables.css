/* SPDX-License-Identifier: Apache-2.0 */
.variables {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

.variables__header h2 {
  margin-bottom: 0.5rem;
}

.variables__header var {
  font-style: normal;
}
