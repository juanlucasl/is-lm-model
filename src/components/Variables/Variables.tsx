/* SPDX-License-Identifier: Apache-2.0 */
import "./Variables.css";
import React from "react";
import VariableInput from "./VariableInput";

export interface Variable {
  state: number;
  setState: (value: number) => void;
  label: string;
}

interface VariablesProps {
  variables: Variable[];
  title?: string;
  description?: string;
}

/**
 * Displays input fields for a set of variables.
 *
 * @returns The input fields for the variables
 */
export const Variables: React.FC<VariablesProps> = ({ variables, title, description }) => (
  <section className="variables">
    {(title || description) && (
      <div className="variables__header">
        {title && <h2>{title}</h2>}
        {description && <var>{description}</var>}
      </div>
    )}
    {variables.map(({ state, setState, label }) => (
      <VariableInput key={label} label={label} step={5} value={state} setValue={setState} />
    ))}
  </section>
);

export default Variables;
