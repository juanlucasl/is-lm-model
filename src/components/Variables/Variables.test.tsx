import React from "react";
import { render, screen } from "@testing-library/react";
import Variables from "./Variables";
import { Variable } from "../../models";

const mockVariables: Variable[] = [
  {
    label: "Mock Variable Input 1",
    value: 0,
    setValue: jest.fn(),
    increaseDescription: "Mock increase description 1",
    decreaseDescription: "Mock decrease description 1",
  },
  {
    label: "Mock Variable Input 2",
    value: 0,
    setValue: jest.fn(),
    increaseDescription: "Mock increase description 2",
    decreaseDescription: "Mock decrease description 2",
  },
];

describe("Variables", () => {
  it("renders correctly", () => {
    const { container } = render(<Variables variables={mockVariables} />);
    expect(container).toMatchSnapshot();
  });

  it("renders the title and description if provided", () => {
    render(
      <Variables variables={mockVariables} title="Mock Title" description="Mock Description" />
    );
    expect(screen.getByRole("heading", { name: /Mock Title/ })).toBeVisible();
    expect(screen.getByText("Mock Description")).toBeVisible();
  });

  it("renders the variable input fields", () => {
    render(<Variables variables={mockVariables} />);
    expect(screen.getByText(mockVariables[0].label)).toBeVisible();
    expect(screen.getByText(mockVariables[1].label)).toBeVisible();
    expect(screen.getAllByTestId("arrow-increase")).toHaveLength(mockVariables.length);
    expect(screen.getAllByTestId("arrow-decrease")).toHaveLength(mockVariables.length);
  });
});
