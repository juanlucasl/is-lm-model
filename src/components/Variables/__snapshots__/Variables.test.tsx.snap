// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Variables renders correctly 1`] = `
<div>
  <section
    class="variables"
  >
    <div
      class="variable-input"
    >
      <span>
        Mock Variable Input 1
      </span>
      <div
        aria-disabled="false"
        aria-label="+ Mock Variable Input 1"
        class="arrow arrow__upwards "
        data-testid="arrow-increase"
        data-tooltip-html="Mock increase description 1"
        id="mockvariableinput1-increase"
        role="button"
      >
        ↑
      </div>
      <div
        aria-disabled="false"
        aria-label="- Mock Variable Input 1"
        class="arrow arrow__downwards "
        data-testid="arrow-decrease"
        data-tooltip-html="Mock decrease description 1"
        id="mockvariableinput1-decrease"
        role="button"
      >
        ↓
      </div>
    </div>
    <div
      class="variable-input"
    >
      <span>
        Mock Variable Input 2
      </span>
      <div
        aria-disabled="false"
        aria-label="+ Mock Variable Input 2"
        class="arrow arrow__upwards "
        data-testid="arrow-increase"
        data-tooltip-html="Mock increase description 2"
        id="mockvariableinput2-increase"
        role="button"
      >
        ↑
      </div>
      <div
        aria-disabled="false"
        aria-label="- Mock Variable Input 2"
        class="arrow arrow__downwards "
        data-testid="arrow-decrease"
        data-tooltip-html="Mock decrease description 2"
        id="mockvariableinput2-decrease"
        role="button"
      >
        ↓
      </div>
    </div>
  </section>
</div>
`;
